Fix voucher offers API endpoint

marjev opened the pull request.

edx-webhook commented:
Thanks for the pull request, @marjev! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fecommerce&number=819

mjfrey commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,118 @@ def offers(self, request):
+            course_ids = [product.course_id for product in products]
+            courses = Course.objects.filter(id__in=course_ids)
+            stock_records = StockRecord.objects.filter(product__in=products)
+            contains_verified_course = 'verified' == benefit.range.course_seat_types
```
this could be a list.  Will this logic work?

marjev replied:
This is always comma-separated string. That's how it's saved in the database.

mjfrey replied:
ok so == will work if the string is "verified, professional"?

marjev replied:
it should be False, in that case. True is returned when offers contains only verified courses.

mjfrey replied:
👍

mjfrey commented:
@mattdrayer please take another look at this.

vkaracic commented on ecommerce/extensions/api/v2/tests/views/test_vouchers.py:
```diff
@@ -6,23 +6,24 @@
 from ecommerce.extensions.catalogue.tests.mixins import CourseCatalogTestMixin
-from ecommerce.extensions.partner.models import StockRecord
 from ecommerce.extensions.test.factories import prepare_voucher
 from ecommerce.coupons.tests.mixins import CouponMixin
```
Can you move this one too, please?

marjev replied:
Why? It's used in `VoucherViewOffersEndpointTests`.

vkaracic replied:
Move up, as in C comes before E :)

mattdrayer commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,120 @@ def offers(self, request):
         query = voucher.offers.first().benefit.range.catalog_query
-        cache_key = 'catalog_query_{}'.format(query)
+        if query:
+            cache_key = 'catalog_query_{}'.format(query)
```
I think it's important that the cache key prefix reflect the information actually being stored in the cache, which in this case is voucher offers.  If we prefix the cache key with "catalog_query", really there could be anything stored in the cache and another workflow could run into problems if it attempts to use the same key prefix in a different context.  So let's go with 'voucher_offers_' for both conditions here.

marjev replied:
Done.

mattdrayer commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,120 @@ def offers(self, request):
                 logger.error('Could not get course information.')
                 return Response(status=status.HTTP_400_BAD_REQUEST)
+            except Http404:
+                logger.error('Could not get product information.')
```
Can we add more context to the log message here?

marjev replied:
Done. I added product title to provide more context, but since this is an edge case that should never happen and in case it happens, this logger should help us indicate that data in DB got corrupted.

mattdrayer commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,120 @@ def offers(self, request):
-
-        offers = []
-        for product in products:
+            )['results']
```
Is it possible that no results key will be available in the response?

marjev replied:
It's a paginated API endpoint, so results will be empty or contains course run info in a form of a list.

mattdrayer commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,120 @@ def offers(self, request):
+                    stock_record = stock_records.get(product__id=product.id)
+                except StockRecord.DoesNotExist:
+                    stock_record = None
+                    logger.error('Stock Record for a product %s not found.', product.id)
```
"...for product %s..."

marjev replied:
Done.

mattdrayer commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,120 @@ def offers(self, request):
-            offers = self._get_offers_for_product(product, voucher, course_info, offers)
+            course_info = get_course_info_from_lms(course_id)
+            stock_record = get_object_or_404(StockRecord, product__id=product.id)
+            course = get_object_or_404(Course, id=course_id)
```
I think it would be better to perform this database lookup before making the network call to the LMS for course_info above

marjev replied:
Done.

mattdrayer commented on ecommerce/extensions/api/v2/views/vouchers.py:
```diff
@@ -72,118 +76,120 @@ def offers(self, request):
+            stock_record = get_object_or_404(StockRecord, product__id=product.id)
+            course = get_object_or_404(Course, id=course_id)
+
+            if course_info and course:
```
if there is no course, will the value be None or will the 404 condition be triggered?

marjev replied:
Done.

mattdrayer commented:
@marjev looking good!  I left some comments and also it looks like the branch needs a rebase -- let me know when you're ready for a final pass.

marjev commented:
@mattdrayer I addressed the comments you left in the previous review. Can you please take another look? Thanks!

Let me know if you have any suggestions or questions.

mattdrayer commented:
changes LGTM -- nice!  👍
